chemprop/uncertainty/uncertainty_calibrator.py

```diff
@@ -733,6 +730,261 @@ def nll(self, preds: List[List[float]], unc: List[List[float]], targets: List[Li
+        uncal_preds = np.array(uncal_predictor.get_uncal_preds())  # shape(data, task)
+        scores = self.nonconformity_scores(uncal_preds)
+
+        cal_preds_in = (scores <= self.tin).astype(int)
+        cal_preds_out = (scores <= self.tout).astype(int)
+        cal_preds = np.concatenate((cal_preds_in, cal_preds_out), axis=1)
+
+        return uncal_preds.tolist(), cal_preds.tolist()
+
+
+class ConformalRegressionCalibrator(UncertaintyCalibrator):
+    """
+    Conformal Calibrator for regression datasets. Used for both conformal regression and conformal quantile regression.
+    Outputs interval of variable size, centered around quantile outputs of model, for each datapoint. Intervals 
+    should cover 1-alpha proportion of datapoints.
+    As discussed in https://arxiv.org/abs/2107.07511.
+    """
+
+    @property
+    def label(self):
+        #return "conformal_quantile_regression"
+        return f"conformal_{self.conformal_alpha}"
+
+    def raise_argument_errors(self):
+        super().raise_argument_errors()
+        if self.dataset_type != "regression":
+            raise ValueError(
+                "Conformal Quantile Regression is only implemented for regression dataset types."
+            )
+
+    def nll(
+        self,
+        preds: List[List[float]],
+        unc: List[List[float]],
+        targets: List[List[float]],
+        mask: List[List[bool]],
+    ):
+        raise NotImplementedError
+
+    @staticmethod
+    def get_preds(predictor: UncertaintyPredictor):
+        return np.array(predictor.get_uncal_preds())
+
+    @staticmethod
+    def get_interval(predictor: UncertaintyPredictor):
+        return np.array(predictor.get_uncal_output())
+
+    def calibrate(self):
+        uncal_interval = self.get_interval(self.calibration_predictor)
+        targets = np.array(self.calibration_data.targets(), dtype=float)  # shape(data, tasks)
+        mask = np.array(self.calibration_data.mask(), dtype=bool)
+
+        self.num_tasks = uncal_interval.shape[1] // 2
+        self.qhats = []
+
+        for task_id in range(self.num_tasks):
+            targets_task_id = targets[:, task_id]
+            targets_task_id = targets_task_id[mask[:, task_id]]
+            uncal_interval_lower = uncal_interval[mask[:, task_id], task_id]
+            uncal_interval_upper = uncal_interval[mask[:, task_id], task_id + self.num_tasks]
+
+            calibration_scores = np.maximum(
+                uncal_interval_lower - targets_task_id, targets_task_id - uncal_interval_upper
+            )
+            calibration_scores = np.append(calibration_scores, np.inf)
+            calibration_scores = np.sort(np.absolute(calibration_scores))
+            self.qhats.append(np.quantile(calibration_scores, 1 - self.conformal_alpha / self.num_tasks, method="higher"))
+
+    def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        uncal_preds = self.get_preds(uncal_predictor)  # shape(data, 2 * task)
```

Comment:
`get_pred()` is not used here? If so, maybe we should remove this L972 and the staticmethod of get_preds(). Besides, I am curious about the shape size of uncal_interval and uncal_preds here. Please see the further comment in the next review comment.

Reply:
We return uncal_preds at the end in order to fit the format of the class of returning the base predictions and the uncertainty output. The shape of uncal_preds should actually be data, tasks and the shape of uncal_intervals is data, 2*tasks